Flink: Maintenance - RewriteDataFiles

pvary opened the pull request:
Compaction implementation.

For details see:
https://docs.google.com/document/d/16g3vR18mVBy8jbFaLjf2JwAANuYOmIwr15yDDxovdnA/edit#heading=h.yd2vbtnf7z6w

pvary commented:
@aokolnychyi, @RussellSpitzer, @szehon-ho - I have touched the Spark compaction as well to remove duplicate code, and move it to the `core` module. Could you please check the Spark, and the core changes?

@stevenzwu: Could you please review the PR?

RussellSpitzer commented:
Could we break this up into smaller PR's? I think just one for the refactoring of classes out of the Spark Module would be a good first step?

pvary commented:
> Could we break this up into smaller PR's? I think just one for the refactoring of classes out of the Spark Module would be a good first step?

Sure thing. I think for these refactors, it is important to see the whole picture as well, and it is always a tough decision to decide how it is easier to review.

Will create the PR soon.

mxm commented:
IMHO just grouping into several commits would suffice, e.g. Core/Spark/Flink. Grouping into multiple PRs, might make this harder to test and review.

ygrzjh commented:
Using the parameter io-impl: `org.apache.iceberg.aws.s3.S3FileIO`, a Kryo serialization exception occurred in the `DataFileRewritePlanner$PlannedGroup` class. The exception details are as follows:

```java
com.esotericsoftware.kryo.KryoException: java.lang.NullPointerException
Serialization trace:
s3 (org.apache.iceberg.aws.s3.S3FileIO)
io (org.apache.iceberg.SerializableTable)
table (org.apache.iceberg.flink.maintenance.operator.DataFileRewritePlanner$PlannedGroup)
	at com.esotericsoftware.kryo.serializers.ObjectField.write(ObjectField.java:82)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:495)
	at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:523)
	at com.esotericsoftware.kryo.serializers.ObjectField.write(ObjectField.java:61)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:495)
	at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:523)
	at com.esotericsoftware.kryo.serializers.ObjectField.write(ObjectField.java:61)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:495)
	at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:599)
	at org.apache.flink.api.java.typeutils.runtime.kryo.KryoSerializer.serialize(KryoSerializer.java:356)
......
```

stevenzwu commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java:
```diff
@@ -0,0 +1,228 @@
+        }
+      }
+
+      int groupsPerCommit =
```
what's the purpose of this calculated number? is it only meaningful when partial progress is enabled?

also the calculation seems incorrect if `maxRewriteBytes` filtered out some groups.

pvary replied:
> what's the purpose of this calculated number? is it only meaningful when partial progress is enabled?

This information is needed by the `DataFileRewriteCommitter` to decide when to commit (if partial commits are enabled)

> also the calculation seems incorrect if maxRewriteBytes filtered out some groups.

Fixed

stevenzwu commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java:
```diff
@@ -0,0 +1,228 @@
+            ctx.timestamp());
+        out.collect(
+            new PlannedGroup(
+                table, groupsPerCommit, rewriter.splitSize(group.sizeInBytes()), group));
```
why do we need to calculate split size here via `rewriter.splitSize()`? target split size is needed for scan planning. this is already after planning.

pvary replied:
This is for calculating the target file size for the rewriter. Might change after the core refactor PR. We will have a place to store it in the plan

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java:
```diff
@@ -40,17 +41,17 @@ public class DeleteFilesProcessor extends AbstractStreamOperator<Void>
   private transient Counter succeededCounter;
 
-  public DeleteFilesProcessor(int taskIndex, String taskName, Table table, int batchSize) {
+  public DeleteFilesProcessor(Table table, String taskName, int taskIndex, int batchSize) {
```
There are multiple places where we have these method parameters. Since we are in the early stages, I have decided to standardize the order. Starting from the widest going to the narrowest filter

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/api/RewriteDataFiles.java:
```diff
@@ -0,0 +1,232 @@
+     *
+     * @param newMaxRewriteBytes to limit the size of the rewrites
+     */
+    public Builder maxRewriteBytes(long newMaxRewriteBytes) {
```
Found it nice to have in Flink to maximize the resources used during a compaction round. Spark currently doesn't have it, and the PR is big enough as it is. If the Spark team finds this useful, we could move this to the common part of the code. In the meantime, my main goal was to keep the Spark part intact

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,304 @@
+            .getOperatorStateStore()
+            .getListState(
+                new ListStateDescriptor<>(
+                    "data-file-updater-in-progress", TypeInformation.of(RewriteFileGroup.class)));
```
Currently using Kryo. The groups are quite complicated objects with Scans and ContentFiles. This PR doesn't optimize this part. It could be added later.

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,304 @@
+    public void commitFileGroups(Set<RewriteFileGroup> fileGroups) {
+      super.commitFileGroups(fileGroups);
+      LOG.debug(
+          "Committed {} for table {} with {}[{}] at {}",
```
Interesting question. Currently the end of the log lines are standard. So it is easy to filter out the relevant lines. We might move the common part to the beginning of the line, but then it is harder to find important information as the lines often truncated. What are your preferences?

pvary replied:
Created a util method for log messages:
```
  public static String format(
      String tableName, String taskName, int taskIndex, long timestamp, String message) {
    return String.format(
        "[For table %s with %s[%d] at %d ]: %s",
        tableName, taskName, taskIndex, timestamp, message);
  }
```

pvary replied:
Needed to change to a constant because of `Slf4jConstantLogMessage`
```
slf4j log statement uses a non-constant expressio
```

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteExecutor.java:
```diff
@@ -0,0 +1,257 @@
+      LOG.debug(
+          "Rewriting files {} from {} for table {} with {}[{}] at {}",
+          value.group().info(),
+          value.group().rewrittenFiles(),
```
Please see my comment about the other log formatting comment.

I would prefer a standard log formatting method, or using MDC.

What are your preferences?

stevenzwu replied:
I would also prefer a standard log formatting. my point is that the log would be harder to read if the middle of the log line is a long arg/variable string. the long arg string can be put in the end of the log line.

e.g. https://github.com/apache/iceberg/blob/6501d29b2d46c8d57f46ad646e2daf7d8865f646/core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java#L69

pvary replied:
Added standard log formatting

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java:
```diff
@@ -0,0 +1,228 @@
+    private final RewriteFileGroup group;
+
+    private PlannedGroup(
+        SerializableTable table, int groupsPerCommit, long splitSize, RewriteFileGroup group) {
```
`DataFileRewriteExecutor` input is rebalanced. It is not trivial to make sure that the table is sent to every instance.

The size of the serialized table is mostly dependent on the size of the FileIO.

stevenzwu replied:
pass `SerializableTable` to the constructor of `DataFileRewriteExecutor` so that it becomes part of the class state.

pvary replied:
We need the current schema/spec for the table. If we pass the data at the constructor then we will work from old table metadata.

pvary replied:
We need the info on the `DataFileRewriteExecutor` as well. That would mean Catalog connection/request from every writer.

TBH, I'm not too concerned about state compatibility. We only has to be very careful about silent issues. In other cases we could always drop the state and start the compaction again. We will not have any data loss if a maintenance task does not finish.

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java:
```diff
@@ -0,0 +1,228 @@
+      }
+
+      RewritePlanResult plan =
+          planner.plan(table, Expressions.alwaysTrue(), table.currentSnapshot().snapshotId());
```
`startingSnapshotId` is the snapshot when the compaction started (files from newer snapshots are ignored during this run). This is the snapshotId which will be used for the new files generated by the compaction.
The name is coming from the existing code, and I tried to avoid modifying core/Spark code as much as possible

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java:
```diff
@@ -0,0 +1,228 @@
+      ListIterator<RewriteFileGroup> iter = groups.listIterator();
+      while (iter.hasNext()) {
+        RewriteFileGroup group = iter.next();
+        if (rewriteBytes + group.sizeInBytes() > maxRewriteBytes) {
```
This would modify existing Spark behavior. We could do this, but I wanted to limit the scope of the changes.

> I also feel it doesn't to be precise to scan the whole list/stream to see if other groups might fit in

Could you please elaborate what do you think here?

stevenzwu replied:
sorry, I meant ` it doesn't need to be precise`. but never mind. we are deleting unneeded items here, which would require iterating through the whole list.

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/api/RewriteDataFiles.java:
```diff
@@ -0,0 +1,232 @@
+ * Creates the data file rewriter data stream. Which runs a single iteration of the task for every
+ * {@link Trigger} event.
+ *
+ * <p>The input is a {@link DataStream} with {@link Trigger} events and every event should be
```
Done

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,304 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * Commits the compaction changes using {@link RewriteDataFilesCommitManager}. The input is a {@link
```
changed compaction -> rewrite

> similarly, move the input/output comment to the append method?

This is an operator so we don't have an append. Shall we add javadoc to the process method?
For me having an operator level javadoc seems better, but if you strongly feel that we need to move the comment, I'm fine with that too.

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,304 @@
+    }
+
+    inProgressState.clear();
+    for (RewriteFileGroup group : inProgress) {
```
`inProgress` is a `Set`, and `addAll` is only working for lists.
`RewriteDataFilesCommitManager` expects a `Set` in `commitFileGroups`, so I used Set for `inProgress`

stevenzwu replied:
not sure if I fully understand you. but `Set` supports `addAll`.
https://docs.oracle.com/javase/8/docs/api/java/util/Set.html#addAll-java.util.Collection-

pvary replied:
`ListStated` doesn't support `addAll(Set)`

stevenzwu replied:
oh. I see. `ListState.addAll(List<>)` only takes in a list (not a set in this case)

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,304 @@
+  @Override
+  public void close() throws IOException {
+    if (commitService != null) {
+      commitService.close();
```
There is no way to guarantee the consistency between the `inProgress` list and the data which is really committed to the Iceberg table. There could be an "old savepoint" which we are restoring, or there could be some delay between storing the state and stopping the job.

If we try to update the table and replace already replaced files, then the commit will fail. So we will not corrupt the table even in these cases.

Given these constraints I would prefer committing everything which is already done.

stevenzwu replied:
> If we try to update the table and replace already replaced files, then the commit will fail. 

so first commit (with new rewrite group offered) after restore would always fail? if yes, that also doesn't seem ideal

> There is no way to guarantee the consistency between the inProgress list and the data which is really committed to the Iceberg table

wondering if commit should only happen in `snapshotState` with partial commit always disabled?

pvary replied:
We can't commit only on checkpoints. We need commit on watermarks too, so upcoming tasks will not conflict with pending commits waiting for checkpoint.

If we accept the statement above, that means that we still have errors on recovery when a watermark arrives on the middle of a checkpoint. On recovery the groups will be compacted again, and we will try to commit them and fail.

So, I don't see an error free solution. This is why I prefer to keep the partial commit feature, and live with the errors on recovery.

Maybe we don't increase the error counters during the first checkpoint? This seems like a bit hacky to me, but might help when monitoring a job

stevenzwu replied:
> We can't commit only on checkpoints. We need commit on watermarks too, so upcoming tasks will not conflict with pending commits waiting for checkpoint.

can we add some metadata into snapshot summary to indicate which trigger has been committed. Then during restore, the committer can check if any restored groups has been committed already. similar to the Flink Iceberg sink using checkpointId to check the commit point. this may require disabling of partial commit, which is probably not necessary with Flink streaming jobs.

pvary replied:
What we need to do: Basically we would need to check every group if it was committed or not. This would mean to keep an up-to-date cache for the already committed groupIds (timestamp + globalId). This also could become "corrupt" if the grouping algorithm or configuration changes.

What we gain: without this we try and validation would prevent duplicate commits. This would happen only during restarts where unaligned checkpointing is used.

I'm starting to get convinced to basically just remove the state from the committer...
When "close" is called on the `CommitService` then every pending group is committed anyways. If we lose some groups, then we lose them... not a big deal. On job restart or on the next compaction they will be regenerated anyways.

Seems like it is less error prone if we just forget about the pending commits and try to commit every group which we receive.

stevenzwu replied:
sure, I am fine with removing the state from the committer

pvary replied:
Removed

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,304 @@
+
+  private CommitService createCommitService(
+      DataFileRewriteExecutor.ExecutedGroup element, long timestamp) {
+    table.refresh();
```
Removed

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,304 @@
+    }
+
+    // Cleanup
+    commitService = null;
```
I delegate this decision to the `CommitService`

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,304 @@
+          timestamp);
+      updateMetrics(fileGroups);
+      inProgress.removeAll(fileGroups);
+      LOG.debug(
```
Theoretically a parallel processor could add more element to the list in the meantime. (The `CommitManager` uses its own thread for commits)

pvary replied:
Changed the `inProgress` to concurrent, and removed the log

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,304 @@
+        }
+
+        for (DataFile rewritten : fileGroup.rewrittenFiles()) {
+          Preconditions.checkArgument(
```
If we somehow get DELETE files then we fail...
Not strictly necessary, but I think it might be nice to have

pvary replied:
Removed the check.
Originally added multiple counters for delete files as well, but then I realized that we don't have those. Removed every check for now

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,304 @@
+              FileContent.DATA.equals(rewritten.content()),
+              "%s is not supported for metrics collection",
+              rewritten);
+          removedDataFileNumCounter.inc();
```
removed matches added

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteExecutor.java:
```diff
@@ -0,0 +1,257 @@
+
+  @Override
+  public void open(Configuration parameters) {
+    this.errorCounter =
```
Different groups, as the `taskName` is different

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TaskResultAggregator.java:
```diff
@@ -0,0 +1,154 @@
+            .getListState(
+                new ListStateDescriptor<>(
+                    "result-aggregator-start-time", BasicTypeInfo.LONG_TYPE_INFO));
+    this.exceptionsState =
```
When does the Exception change?
What do you suggest?

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/api/RewriteDataFiles.java:
```diff
@@ -0,0 +1,232 @@
+              .forceNonParallel();
+
+      return trigger
+          .union(updated)
```
For watermarks

pvary replied:
Basically we just reuse the watermark coming from the `DataFileRewriteCommitter` to trigger the watermark for the aggregator

pvary replied:
That one makes sure that the `TaskResultAggregator` watermark is not triggered until all of the compaction groups are committed

stevenzwu replied:
> That one makes sure that the TaskResultAggregator watermark is not triggered until all of the compaction groups are committed

that would need the `updated` stream. but why do we need the original `trigger` stream from input arg?

pvary replied:
We get the start time for Task from the original Trigger record

stevenzwu commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/api/RewriteDataFiles.java:
```diff
@@ -0,0 +1,232 @@
+          .union(updated)
+          .connect(
+              planned
+                  .getSideOutput(TaskResultAggregator.ERROR_STREAM)
```
oh. I see. wondering if the output of the planner, executor, committer could be just `Pair` of exception or value. this way, the DAG would be simpler.

pvary replied:
I think the operator codes are easier to understand, and it is better separation of responsibility altogether to keep the sideoutput for the errors.

stevenzwu replied:
sounds good

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,318 @@
+  private void commitInProgress(long timestamp) {
+    if (!inProgress.isEmpty()) {
+      try {
+        FlinkRewriteDataFilesCommitManager manager =
```
The parametrization is different as we want to commit everything at once. Also this is only 2 lines of code, so I skipped reusing the other method

pvary replied:
Ok.. after some refactor, found a way to reuse the code

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,318 @@
+        CommitService service = manager.service(Integer.MAX_VALUE);
+        service.start();
+        manager.commitOrClean(inProgress);
+        service.close();
```
Moved

stevenzwu commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,320 @@
+        }
+      }
+
+      table.refresh();
```
is this refresh needed?

pvary replied:
Removed the refresh, and the snapshotId from the next log message

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,320 @@
+      ++processed;
+    } catch (Exception e) {
+      LOG.info(
+          LogUtil.MESSAGE_PREFIX + "Exception processing {}",
```
This makes sure that we change the message consistently, so I prefer this way

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,320 @@
+      }
+    }
+
+    // Has to be concurrent since it is accessed by the CommitService from another thread
```
It is fine as there are no duplicates generated, and we never have multiple instances of the same group anyway

stevenzwu replied:
In this case, I don't see any difference btw Set and List. Should we just use List to be more clear?

pvary replied:
It ComitService input was a Set, so I have used set to collect the inputs.
Irrelevant now, as we don't collect anything in the state now

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java:
```diff
@@ -0,0 +1,230 @@
+        taskIndex,
+        ctx.timestamp());
+    try {
+      SerializableTable table =
```
Refresh needs to turn to the catalog to ensure that there is no newer version of the table.
So this is still a Catalog request for every Executor

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java:
```diff
@@ -0,0 +1,230 @@
+      }
+
+      RewritePlanResult plan =
+          planner.plan(
```
That's a possibility for later optimization. It is a non-trivial task to keep a table state in memory and keep it sync with the actual table - especially when there are independent writers to the table

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java:
```diff
@@ -0,0 +1,230 @@
+      }
+    } catch (Exception e) {
+      LOG.info(
+          LogUtil.MESSAGE_PREFIX + "Exception planning data file rewrite groups",
```
Done

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteExecutor.java:
```diff
@@ -0,0 +1,257 @@
+          taskName,
+          taskIndex,
+          ctx.timestamp(),
+          value.group().rewrittenFiles().size());
```
Added

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteExecutor.java:
```diff
@@ -0,0 +1,257 @@
+  }
+
+  private TaskWriter<RowData> writerFor(PlannedGroup value) {
+    String formatString =
```
The goal of the compaction is to reorganize the table data to match the `TableProperties`. I don't think it is a good idea to allow overwriting those.

If the need raises later, then we could still decide to do so after enhancing the API to get the extra configurations for the writeOptions.

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteExecutor.java:
```diff
@@ -0,0 +1,257 @@
+            value.table().properties(),
+            null,
+            false);
+    factory.initialize(subTaskId, attemptId);
```
This is the question of how flexible we want the compaction flow to be.
Creating the factory once during the open method would mean, that the compaction will always write with the old schema - even if the table is updated in the meantime

pvary replied:
MonitorSource provides the possibility to have a standalone job to run the maintenance for a given table.
While changing the table schema/spec behind a streaming ingest job will not cause issues, if we start compacting new files with the old schema we will start losing data from the new colums, as the will not be written to the compacted files. This should be prevented by either adding guards against it, or by making sure we are able to handle dynamic changes.
Discussion with users around the DynamicSink convinced me, that this is a valid concern, and we should move forward with dynamic solutions whenever it is possible.

@mxm: What are your thoughts?

mxm replied:
We could always make this configurable, but I would suggest to default to the dynamic variant. The reason mainly being that it doesn't hurt to use the up-to-date table metadata with the current Iceberg sink. Plus, it enables to run in standalone mode, where we definitely want to refresh the table metadata. Note that being static with Flink is much worse than with Spark, because of the long-running nature of Flink streaming applications. Once we attach the compaction to the Dynamic Sink, not being dynamic is out of question.

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteExecutor.java:
```diff
@@ -0,0 +1,257 @@
+            value.table().schema(),
+            value.table().schema(),
+            PropertyUtil.propertyAsString(value.table().properties(), DEFAULT_NAME_MAPPING, null),
+            false,
```
I don't get your question here.
The `RowDataFileScanTaskReader` is an existing class with the existing `caseSensitive` feature.

Are you suggesting, that we should create a new parameter for the compaction flow, like `caseSensitive`? Given that there were recent changes in the Spark flow wrt case sensitivity in file filtering, I think we can do this here as well.

pvary replied:
We can add a new config to set this, but I don't see too much value.
When would this be important?

stevenzwu replied:
sorry for the confusion. I meant if it is right to hard `caseSensitive` to `false`. fixed the original comment.

I guess it is not important right now since there is no filtering based on column names.

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteExecutor.java:
```diff
@@ -0,0 +1,257 @@
+          subTaskId,
+          attemptId);
+      writer.abort();
+      LOG.info(
```
Removed

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,320 @@
+        service = createCommitService(timestamp, inProgress.size());
+        inProgress.forEach(service::offer);
+      } catch (Exception e) {
+        LOG.info(
```
Moved it to warn.
The process could still survive the error, so I think it is enough to use warn

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,320 @@
+          try {
+            service.close();
+          } catch (Exception e) {
+            LOG.warn(
```
Moved it to warn. Same reasoning as above

pvary commented:
To summarise the state, I see the following open questions:
- State handling and restoration
- How to handle table schema, specification, configuration changes
- How to handle errors

**State handling and restoration**
What we know:
- There will be no issues with aligned checkpoints, and the checkpoint barriers are aligned with the triggers, so every event is processed in a single checkpoint
- With unaligned checkpoints
    - The state of the compaction job is not super important as the user could always trigger a new compaction and drop old pending results
    - There are good barriers against double committing the same compaction results. If a replaced file doesn’t exists in the table anymore then the commit will fail
    - If we want to use the core CommitManager strategy (which handles partial commits), then we can’t guarantee the consistency of the state and the Iceberg table, since the CommitManager uses its own thread to do the actual commits - here the Flink folks might come up with some solution currently unknown to me
    - If we restore an old checkpoint or savepoint we are almost guaranteed to fail to commit the inProgress records as they are most probably already committed by the previous iteration of the job

Currently the following ideas emerged to handle the situation:
- Use core CommitManager, store the state and do a best effort when committing the inProgress records
    - Helps if the job stopped ungracefully (without calling the close on the DataFileRewriteCommitter
    - Generates errors on start (gracefully handled)
    - Generates errors on receiving groups stored in the wire state (gracefully handled)
- Create our own strategy for committing
    - Needs an elaborate strategy to store the group ids in the commits
    - Before adding a new group we need to scan the table if the group is already committed - either with cache, or continuous checks
    - We lose the features provided by the core CommitManager (groups handling and parallel commits)
- Remove state from the Compaction Task
    - Simplest implementation
    - We might lose some in flight work - nothing on graceful stop, but there might be some loss on crashing jobs

**How to handle table schema, specification, configuration changes**
We could decide if we want to handle table schema/specification/configuration changes.
- We could opt for foregoing those changes and use the values available when the job started (Like every current Flink job) 
- We need to add guardrails to prevent compaction when there are some changes in the table which could cause issues. Like losing date from new columns when reading new files with the old schema
- We could opt for keeping the schema/specification/configuration up-to-date. For this we have the following options:
    - Generate a SerializableTable object at planning time, and send it through the wire for the DataFileRewriteExecutors
        - The size of the table could increase the network traffic
        - There is some issue with S3FileIO serialization which needs to be fixed
    - Send only the snapshotId to the DataFileRewriteExecutors and load the table on every executor
        - Adds extra load to the Catalog, as every executor needs to fetch the table data once for every trigger

**How to handle errors**
We have 2 options for propagating the errors in the specific operators
- Use side output to emitt the error and aggregate it in an final operator
    - Better separation of concerns
- Send a Pair<Data, Exception> as an output and handle the exceptions in every operator
    - Simpler flow

Another ortogonal question wrt. the exceptions is what to propagate:
- Propagate the whole exception
    - We have the full stack trace available at the aggregation point
    - Might cause state issues if the Exception is changed
- Propagate the exception message only and log the exception locally
    - We have less network traffic

My current preferences:
- **State handling and restoration**: 3
- **How to handle table schema, specification, configuration changes**: 2/a
- **How to handle errors**: 1, 1

pvary commented:
Added a PR to remove the state handling (we can still revert if we decide so)

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LogUtil.java:
```diff
@@ -0,0 +1,26 @@
+ */
+package org.apache.iceberg.flink.maintenance.operator;
+
+class LogUtil {
+  static final String MESSAGE_PREFIX = "[For table {} with {}[{}] at {}]: ";
+  static final String MESSAGE_FORMAT_PREFIX = "[For table %s with {%s}[{%d}] at {%d}]: ";
+
+  private LogUtil() {}
```
They will be used in multiple operators.
Any better ideas?

mxm replied:
Yes, I suggesting to put them into DataFileRewritePlanner as public constants. At least, this puts them in context of the DataFileRewrite operators. `LogUtil` is super generic.

pvary replied:
I did so, but I feel we will find it strange to have the constants in the DataFileRewritePlanner when we start using it in operators for entirely different Tasks

mxm replied:
Ok, not important.

pvary replied:
If we don't use a constant, the we need to check every comment line if all of the relevant parameters are logged. I have tried several solutions, but this one seemed the best for me. Notice the constant is used many times

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/api/TestRewriteDataFiles.java:
```diff
@@ -0,0 +1,447 @@
+  }
+
+  @Test
+  void testStateRestore(@TempDir File savepointDir) throws Exception {
```
You are right
 I will remove as soon as I have my new machine setup [](url)

pvary replied:
Removed

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/api/TableMaintenance.java:
```diff
@@ -83,7 +83,7 @@ public static Builder forChangeStream(
   /**
    * Use this for standalone maintenance job. It creates a monitor source that detect table changes
-   * and build the maintenance pipelines afterwards.
+   * and build the maintenance pipelines afterward.
```
Reverted

mxm commented:
@pvary @stevenzwu Could we wrap this up? I don't see any fundamental issues here. The only open point from my perspective is table refresh during planning time, but even that could be handled in a follow-up.

pvary commented:
> @pvary @stevenzwu Could we wrap this up? I don't see any fundamental issues here. The only open point from my perspective is table refresh during planning time, but even that could be handled in a follow-up.

@mxm: We decided to refactor out #11513 to make this work. Yesterday I have discussed this with Russell, and I will post a message on the dev list about the main remaining question (https://github.com/apache/iceberg/pull/11513#discussion_r1877424024). We need to decide on how the future API will look like.

We would like to build the Flink Maintenance above the new API since the current code would need serious code duplication.

pvary commented:
Rebased and changed to use the new API (#12306) and implementation (#12493)

pvary commented:
Rebased based on the new API.
@stevenzwu: Could you please review?

Thanks,
Peter

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java:
```diff
@@ -0,0 +1,209 @@
+      SerializableTable table =
+          (SerializableTable) SerializableTable.copyOf(tableLoader.loadTable());
+      if (table.currentSnapshot() == null) {
+        LOG.info(
```
We have different number of parameters for different uses, so the simple log method is not an option. Also exception handling, and debug level messages are needed to be handled.
We could have a method for `DataFileRewritePlanner.MESSAGE_PREFIX + message`, but that's not worth it IMHO.

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDataFileRewritePlanner.java:
```diff
@@ -0,0 +1,196 @@
+    assertThat(planWithNoLimit).hasSize(2);
+
+    // Second run with limit
+    long limit =
```
`maxRewriteBytes` is set to `limit`, so this prevents adding any new groups to this compaction during the planning phase. The test case is for testing the `rewriteBytes` functionality

stevenzwu replied:
previous `planWithNoLimit` with 10 MB limit has two groups planned. this `limit` is smaller and planner would generate one group. this is the part I didn't quite follow.

pvary replied:
When the plan reaches the limit, then no consecutive group should be added to the plan.

The `maxRewriteBytes` is added when we found not maintained tables, and we wanted to start running compaction on them. In this case, the first run is a whole table rewrite. If the table is big enough we can face resource issues. To solution was to run the compaction for limited data only. With the limit the user can configure a maximum size for the task, and consecutive runs incrementally could compact the remaining (not yet compacted) files.

Added some comments to the test.

stevenzwu replied:
got it. it is `maxRewriteBytes` limit. earlier I thought it is the rewrite group limit.

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDataFileRewriteRunner.java:
```diff
@@ -0,0 +1,351 @@
+  @Test
+  void testV2Table() throws Exception {
+    Table table = createTableWithDelete();
+    update(table, 1, null, "a", "b");
```
In an upsert stream, there might not be an actual record

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDataFileRewriteRunner.java:
```diff
@@ -0,0 +1,351 @@
+
+      assertThat(testHarness.getSideOutput(TaskResultAggregator.ERROR_STREAM)).isNull();
+      planned = testHarness.extractOutputValues();
+      assertThat(planned).hasSize(1);
```
This is because the `limit` is set

stevenzwu replied:
I know `limit` is set to the size of the first 2 files. but there are 4 files in the table, shouldn't planner generate at least 2 groups? I am missing the connection here. similar to another comment.

pvary replied:
This is a different `limit`. Renamed the variables to help readability.

This test is about testing the target file size for the runner.

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/api/RewriteDataFiles.java:
```diff
@@ -0,0 +1,235 @@
+    private long maxRewriteBytes = Long.MAX_VALUE;
+
+    /**
+     * Allows committing compacted data files in batches. For more details description see {@link
```
Fixed here, and all of the other places

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/api/RewriteDataFiles.java:
```diff
@@ -0,0 +1,235 @@
+    }
+
+    /**
+     * Every other option is overridden, and all the files are rewritten.
```
Fixed

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/api/RewriteDataFiles.java:
```diff
@@ -0,0 +1,235 @@
+
+    /**
+     * The input is a {@link DataStream} with {@link Trigger} events and every event should be
+     * immediately followed by a {@link org.apache.flink.streaming.api.watermark.Watermark} with the
```
Done

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/api/TableMaintenance.java:
```diff
@@ -300,8 +300,7 @@ private DataStream<TableChange> changeStream(String tableName, TableLoader loade
     private static String nameFor(MaintenanceTaskBuilder<?> streamBuilder, int taskIndex) {
-      return String.format(
-          "%s [%s]", streamBuilder.getClass().getSimpleName(), String.valueOf(taskIndex));
+      return String.format("%s [%s]", streamBuilder.getClass().getSimpleName(), taskIndex);
```
Done

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java:
```diff
@@ -0,0 +1,209 @@
+  @Override
+  public void processElement(Trigger value, Context ctx, Collector<PlannedGroup> out)
+      throws Exception {
+    LOG.debug(
```
Done

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java:
```diff
@@ -0,0 +1,209 @@
+          // Keep going, maybe some other group might fit in
+          LOG.info(
+              DataFileRewritePlanner.MESSAGE_PREFIX
+                  + "Skipping group {} as max rewrite size reached",
```
Done

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java:
```diff
@@ -0,0 +1,209 @@
+          groups);
+
+      for (RewriteFileGroup group : groups) {
+        LOG.debug(
```
Done

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewritePlanner.java:
```diff
@@ -0,0 +1,209 @@
+        out.collect(new PlannedGroup(table, groupsPerCommit, group));
+      }
+    } catch (Exception e) {
+      LOG.info(
```
Done

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteRunner.java:
```diff
@@ -0,0 +1,252 @@
+          taskName,
+          taskIndex,
+          ctx.timestamp(),
+          value.group().info(),
```
Done

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteRunner.java:
```diff
@@ -0,0 +1,252 @@
+      throws Exception {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug(
+          DataFileRewritePlanner.MESSAGE_PREFIX + "Rewriting files {} from {}",
```
Done

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,215 @@
+  private class FlinkRewriteDataFilesCommitManager extends RewriteDataFilesCommitManager {
+    private final long timestamp;
+
+    FlinkRewriteDataFilesCommitManager(Table table, long snapshotId, long timestamp) {
```
Done

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,215 @@
+    @Override
+    public void commitFileGroups(Set<RewriteFileGroup> fileGroups) {
+      super.commitFileGroups(fileGroups);
+      LOG.debug(
```
Done

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,215 @@
+      if (commitService != null) {
+        commitService.close();
+        if (processed != commitService.results().size()) {
+          throw new RuntimeException(
```
Done

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,215 @@
+    try {
+      if (commitService != null) {
+        commitService.close();
+        if (processed != commitService.results().size()) {
```
Removed

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/api/MaintenanceTaskInfraExtension.java:
```diff
@@ -48,16 +48,7 @@ class MaintenanceTaskInfraExtension implements BeforeEachCallback {
-            .name(IGNORED_OPERATOR_NAME)
-            .forceNonParallel();
-    sink = new CollectingSink<>();
+    init(StreamExecutionEnvironment.getExecutionEnvironment());
```
Reverted

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/api/MaintenanceTaskTestBase.java:
```diff
@@ -37,16 +39,54 @@ class MaintenanceTaskTestBase extends OperatorTestBase {
+    runAndWaitForResult(env, triggerSource, collectingSink, true, null, () -> true);
+  }
+
+  Configuration runAndWaitForResult(
```
Removed

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/api/TestRewriteDataFiles.java:
```diff
@@ -0,0 +1,409 @@
+    insert(table, 1, "a");
+    insert(table, 2, "b");
+    insert(table, 3, "c");
+    insert(table, 4, "d");
```
Done

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/api/TestRewriteDataFiles.java:
```diff
@@ -0,0 +1,409 @@
+                    DUMMY_TASK_NAME,
+                    "0",
+                    ERROR_COUNTER),
+                1L)
```
Done

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/api/TestRewriteDataFiles.java:
```diff
@@ -0,0 +1,409 @@
+                    DUMMY_TASK_NAME,
+                    "0",
+                    ADDED_DATA_FILE_NUM_METRIC),
+                1L)
```
Done

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/api/TestRewriteDataFiles.java:
```diff
@@ -0,0 +1,409 @@
+
+    runAndWaitForSuccess(infra.env(), infra.source(), infra.sink());
+
+    // Check the metrics
```
Removed

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/OperatorTestBase.java:
```diff
@@ -124,12 +134,62 @@ protected static Table createTableWithDelete() {
+    table.newRowDelta().addRows(dataFile).addDeletes(eqDelete).commit();
+  }
+
+  protected void update(Table table, Integer id, String oldData, String tempData, String newData)
```
Added javadoc

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,221 @@
+  }
+
+  private static void assertDataFiles(
+      Table actual, Set<DataFile> expectedAdded, Set<DataFile> expectedRemoved) {
```
Done

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,221 @@
+
+      testHarness.processElement(rewritten.get(1), EVENT_TIME);
+      // This should be committed synchronously
+      assertDataFiles(
```
If the compaction doesn't happen, then the added/deleted files are different

pvary replied:
The `assertDataFiles` not only checks the numbers, but checks the locations as well. So I kept that.
Added a check to ensure that the groups of the `rewritten` tasks are different, to harden the test cases.

Also added a check for the expected data file in the table

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,221 @@
+          table, rewritten.get(0).group().addedFiles(), rewritten.get(0).group().rewrittenFiles());
+
+      testHarness.processElement(rewritten.get(1), EVENT_TIME);
+      // This should be committed synchronously
```
Removed

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDataFileRewriteCommitter.java:
```diff
@@ -0,0 +1,221 @@
+
+      testHarness.processElement(updateBatchSize(rewritten.get(0)), EVENT_TIME);
+      assertNoChange(table);
+      testHarness.processElement(updateBatchSize(rewritten.get(1)), EVENT_TIME);
```
in the `updateBatchSize` sets the size to 2

pvary replied:
I don't want to add a method to the production code for testing purposes only.
Maybe you have an idea for a better name to make the code more readable?

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDataFileRewritePlanner.java:
```diff
@@ -0,0 +1,196 @@
+  }
+
+  void assertRewriteFileGroup(
+      DataFileRewritePlanner.PlannedGroup actual, Table table, Set<DataFile> files) {
```
done

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDataFileRewritePlanner.java:
```diff
@@ -0,0 +1,196 @@
+            actual.group().fileScanTasks().stream()
+                .map(s -> s.file().location())
+                .collect(Collectors.toSet()))
+        .isEqualTo(files.stream().map(ContentFile::location).collect(Collectors.toSet()));
```
Done

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDataFileRewritePlanner.java:
```diff
@@ -0,0 +1,196 @@
+    FileScanTask withDelete = tasks.get(0).deletes().isEmpty() ? tasks.get(1) : tasks.get(0);
+    assertThat(withDelete.deletes()).hasSize(2);
+    // Find the equality delete and the positional delete
+    if (withDelete.deletes().get(0).content() == FileContent.EQUALITY_DELETES) {
```
The deleteFile equals is not implemented

pvary replied:
Done

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDataFileRewriteRunner.java:
```diff
@@ -0,0 +1,351 @@
+  void assertRewriteFileGroup(
+      DataFileRewriteRunner.ExecutedGroup actual,
+      Table table,
+      Collection<Record> records,
```
Done

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDataFileRewriteRunner.java:
```diff
@@ -0,0 +1,351 @@
+      PartitionData newPartition = new PartitionData(table.spec().partitionType());
+      newPartition.set(0, "p1");
+      newPartition.set(1, 4);
+      PartitionData[] transformedPartitions = {
```
Done

pvary commented:
@stevenzwu: Thanks for the review!
I was OOO for a while, but I'm back now. Addressed all of your concerns. If you have time, could you please review?

Thanks, Peter

Guosmilesmile commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TaskResultAggregator.java:
```diff
@@ -0,0 +1,103 @@
+  private transient Long startTime;
+
+  public TaskResultAggregator(String tableName, String taskName, int taskIndex) {
+    org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkNotNull(
```
Why do we need to reference the Preconditions from both packages?

pvary replied:
Good catch!
Fixed here, but we need another PR to fix everywhere in the Flink code. Also please update the `.baseline/checkstyle/checkstyle.xml` to prevent this class to be imported again"

Guosmilesmile replied:
Thank you for the quick fix! I'll create a PR to get this done.

Guosmilesmile reviewed (commented):
Some small confuse

pvary commented:
Merged to main.
Thanks for the review @stevenzwu, @Guosmilesmile, @RussellSpitzer, @mxm!
